[clang] Fix static local variables in consteval functions

Fznamznon opened the pull request:
Because the initializers for static locals were only emitted once the body
of the parent functions were emitted, we were always missing
initializers for static locals defined inside of consteval functions
because their bodies are never handled by CodeGen. This patch attempts
to fix that by fitting such variables into existing lazy emission
system.

Fixes https://github.com/llvm/llvm-project/issues/82994

Fznamznon committed:
[clang] Emit static const/constexpr variables once they're met

Do not wait until the body of the function is met. This helps with
initializing static variables in conseval functions.

Fixes https://github.com/llvm/llvm-project/issues/82994

rjmccall reviewed (commented):
Initializers can be self-referential, so doing this eagerly might be problematic.

Why does this help with initializing things in consteval functions?

Fznamznon committed:
[clang] Fix static local variables in consteval functions

Because the initializers for static locals were only emitted once the body
of the parent functions were emitted, we were always missing
initializers for static locals defined inside of consteval functions
because their bodies are never handled by CodeGen. This patch attempts
to fix that by fitting such variables into existing lazy emission
system.

Fznamznon committed:
Merge branch 'main' into gh82994

Fznamznon committed:
Check that there is no decl for consteval function

rjmccall commented on clang/lib/CodeGen/CGDecl.cpp:
```diff
@@ -432,7 +443,8 @@ void CodeGenFunction::EmitStaticVarDecl(const VarDecl &D,
   // If this value has an initializer, emit it.
-  if (D.getInit() && !isCudaSharedVar) {
+  if (D.getInit() && !isCudaSharedVar &&
+      !(D.isUsableInConstantExpressions(getContext()) && D.isStaticLocal())) {
```
This function is specific to static local variables, right? So I think `D.isStaticLocal()` is always true.

Can we skip emitting the global variable completely?  What I'd like to see is that we just don't emit any IR associated with the static local at all if `isUsableInConstantExpressions` is true and the variable is never ODR-used. Like, it might be a little more complicated than that in corner cases — we'll have to emit the type in any case if it's variably-modified, and maybe these CUDA shared variables shouldn't be delayed (although maybe they're not usable in constant expressions anyway?) — but there's really no reason to emit a global declaration here that I can see.

rjmccall commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -4461,8 +4461,14 @@ void CodeGenModule::EmitGlobalDefinition(GlobalDecl GD, llvm::GlobalValue *GV) {
 
-  if (const auto *VD = dyn_cast<VarDecl>(D))
+  if (const auto *VD = dyn_cast<VarDecl>(D)) {
+    if (VD->isStaticLocal() && !getContext().shouldExternalize(D)) {
```
What's the deal with `shouldExternalize`?

Fznamznon commented:
> Sorry I missed this PR before - I think John's comment make sense and I don't have much to add - ping me when you need further reviews!

Yeah, sorry for the delay on this, work got in the way. \*dusting the branch\*

Fznamznon committed:
Merge branch 'main' into gh82994

Fznamznon committed:
Merge branch 'main' into gh82994

Fznamznon committed:
Fix stale test
